Fix IOSStorageFolder permission problem

sghctoma opened the pull request:
## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
This PR aims to fix `StorageProvider.OpenFolderPickerAsync` permission errors when run on an iOS device.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
Folders opened using `StorageProvider.OpenFolderPickerAsync` can't be used on an iPhone, because of "Operation not permitted" errors on every operation. 

<img width="1338" alt="OpenFolderPickerAsync-error" src="https://github.com/AvaloniaUI/Avalonia/assets/214533/73ca7893-c7f4-4457-be3c-fd5c79aaf924">

Please note that these errors occur only on a real device, the emulator works without the changes in this PR. My test device runs iOS 17.1.2, and I do not have access to any other Apple devices / iOS versions to test on.

I have attached a [minimal test application](https://github.com/AvaloniaUI/Avalonia/files/13694032/OpenFolderTest.zip) to demonstrate this behavior. The application asks the user for a folder, and puts the name of all files and folders under it into a `ListBox`. The file/folder list will appear in the `ListBox` when run in the emulator, but not when run on a real device.

## What is the updated/expected behavior with this PR?
With the commit in this PR, folders opened using `StorageProvider.OpenFolderPickerAsync` can be accessed as normal.

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->
The root cause of the problem is that `IOSStorageFolder` does not call `StartAccessingSecurityScopedResource` / `StopAccessingSecurityScopedResource` on its `Url`. This PR adds these calls to `IOSStorageFolder`'s constructor and `Dispose` methods, similarly to `IOSSecurityScopedStream`.

sghctoma committed:
Fix IOSStorageFolder permission problem

This commit fixes "Operation not permitted" errors when accessing
folders opened via StorageProvider.OpenFolderPickerAsync by calling
[Start/Stop]AccessingSecurityScopedResource on IOSStorageFolder's
Url.

sghctoma commented:
@dotnet-policy-service agree

maxkatz6 commented on src/iOS/Avalonia.iOS/Storage/IOSStorageItem.cs:
```diff
@@ -150,6 +150,7 @@ internal sealed class IOSStorageFolder : IOSStorageItem, IStorageBookmarkFolder
 {
     public IOSStorageFolder(NSUrl url) : base(url)
     {
+        Url.StartAccessingSecurityScopedResource();
```
I don't think StartAccessingSecurityScopedResource should be global here for a whole lifetime of the folder. Not to mention, Dispose does nothing in this case.

Wouldn't it be better to add StartAccessingSecurityScopedResource + Stop to the specific operations? Like, GetItemsAsync.

sghctoma replied:
Yes, you are right. I was debating this with myself and made the wrong decision.

I started converting to per-operation calls, but I'm not sure how to handle StartAccessingSecurityScopedResource's return value. Should I return early when it fails like IOSStorageItem.SaveBookmarkAsync does, or should I just let it fail later? An argument for the latter is that a permission denied error would be propagated to the caller\* instead of just receiving null.

\* At least for GetItemsAsync and CreateFolderAsync. CreateFileAsync does not throw, because NSFileManager.DefaultManager.CreateFile does not give error information.

maxkatz6 replied:
I don't see any problem with not throwing early here, as iOS native exception might have more detailed information about the error. Or, alternatively, it can be wrapped as an inner exception.

But either way, if StartAccessingSecurityScopedResource returned false, we shouldn't call StopAccessingSecurityScopedResource then.

sghctoma replied:
Thanks, I went with iOS native exceptions. 

[Apple's documentation](https://developer.apple.com/documentation/foundation/nsurl/1413736-stopaccessingsecurityscopedresou) says "_If you call this method on a URL whose referenced resource you don’t have access to, nothing happens_.", so I think it's OK to call StopAccessingSecurityScopedResource even if the Start failed.

I hit a snag with MoveAsync, which led me to other problems with accessing content in opened folders. These all boil down to the fact that only the IOSStorageFolder directly opened with StorageProvider.OpenFolderPickerAsync has a security-scoped NSUrl, the IOSStorageItems acquired via CreateFileAsync, CreateFolderAsync and GetItemsAsync from this folder do not, even though we have access to them (UIDocumentPickerViewController gives recursive access) \*. In practice, this means calling StartAccessingSecurityScopedResource on such objects' Url fails; we need to call it on the security-scoped ancestor's Url. My solution stores this ancestor's Url in every IOSStorageItem, and calls Start/StopAccessingSecurityScopedResource on this instead of it's own.

\* _I'd like to clarify that this is not an Avalonia, or even Xamarin.iOS thing, this behaviour comes from Apple's APIs._

sghctoma committed:
Narrow down security scoped resource access scope

Wrap individual operations in [Start/Stop]AccessingSecurityScopedResource
calls instead of retaining access for the whole lifetime of the folder.

maxkatz6 reviewed (approved):
LGTM
